README.md
# go-hll

[![Latest Version](http://img.shields.io/github/release/mtchavez/go-hll.svg?style=flat-square)](https://github.com/mtchavez/go-hll/releases)
[![Test](https://github.com/mtchavez/go-hll/actions/workflows/test.yml/badge.svg)](https://github.com/mtchavez/go-hll/actions/workflows/test.yml)
[![Go Documentation](http://img.shields.io/badge/go-documentation-blue.svg?style=flat-square)](http://godoc.org/github.com/mtchavez/go-hll)
[![Go Report Card](https://goreportcard.com/badge/github.com/mtchavez/go-hll)](https://goreportcard.com/report/github.com/mtchavez/go-hll)
[![Maintainability](https://api.codeclimate.com/v1/badges/cbdf4f5b5cfa83ad2030/maintainability)](https://codeclimate.com/github/mtchavez/go-hll/maintainability)
[![Test Coverage](https://codecov.io/gh/mtchavez/go-hll/branch/master/graph/badge.svg?token=YBShFgosRF)](https://codecov.io/gh/mtchavez/go-hll)

Go implementation of Hyper Log Log

## Install

```go
go get github.com/mtchavez/go-hll/hll
```

## Usage

- [New](#create-new)
- [New With Default Error](#new-with-default-error)
- [Adding](#adding)
- [Count](#count)

### Create new

New hyper log log table with a desired error

```go
package main

import (
  "github.com/mtchavez/go-hll/hll"
)

func main() {
  hll := NewWithErr(0.065)
}
```

### New with default error

```go
package main

import (
  "github.com/mtchavez/go-hll/hll"
)

func main() {
  // Uses DefaultErr of 0.065
  hll := New()
  hll.Add("foo")
}
```

### Adding

Add some words to the table

```go
package main

import (
  "github.com/mtchavez/go-hll/hll"
)

func main() {
  hll := NewWithErr(0.065)
  words := []string{"apple", "this is bananas", "kiwi kiwi kiwi", "Peach is a peach", "apple banana peach wiki pear"}
  for _, word := range words {
    hll.Add(word)
  }
}
```

### Count

Get the count and calculate the error of your hyper log log

```go
hll := NewWithErr(0.065)
  words := []string{"apple", "this is bananas", "kiwi kiwi kiwi", "Peach is a peach", "apple banana peach wiki pear"}
  for _, word := range words {
    hll.Add(word)
  }
  count := hll.Count()
  err := float32((count - uint32(len(words))) / uint32(len(words)) / 100.0)
  fmt.Printf("\nCount: %d\nError: %f%%\n\n", count, err)

```

## Documentation

Docs can be generated locally using ```godoc``` or go to [godoc.org](http://godoc.org/github.com/mtchavez/go-hll/hll)

## Benchmarks

```txt
# Updated: 2022-07-01

goos: darwin
goarch: arm64
pkg: github.com/mtchavez/go-hll/hll

BenchmarkHllNew-8         103207             11357 ns/op
BenchmarkHllAdd-8       15996489             74.95 ns/op
BenchmarkHllCount-8       156948              7598 ns/op
```

## Tests

Run tests using ```go test```.

## License

Written by Chavez

Released under the MIT License: <http://www.opensource.org/licenses/mit-license.php>


CHANGELOG.md
## 0.1.0 (2022-07-01)

- Tagged release for go mod
- Initial features
  - Adding
  - Count


hll/hll_test.go
package hll

import (
	"fmt"
	"math"
	"testing"
)

func TestGetAlpha(t *testing.T) {
	result := getAlpha(16)
	if result != 0.673 {
		t.Errorf("Expected alpha of 0.673 but got %f", result)
	}

	result = getAlpha(32)
	if result != 0.697 {
		t.Errorf("Expected alpha of 0.697 but got %f", result)
	}

	result = getAlpha(64)
	if result != 0.709 {
		t.Errorf("Expected alpha of 0.709 but got %f", result)
	}

	result = getAlpha(128)
	if fmt.Sprintf("%f", result) != fmt.Sprintf("%f", 0.715270) {
		t.Errorf("Expected alpha of 0.7152705 but got %f", result)
	}
}

func TestNew(t *testing.T) {
	hll := New()
	if hll.mapSize != 256.0 {
		t.Errorf("Map size should be 256 but got %f", hll.mapSize)
	}

	if len(hll.Table) != int(hll.mapSize) {
		t.Errorf("Table size should be %f but got %d", hll.mapSize, len(hll.Table))
	}

	if hll.comp != 24 {
		t.Errorf("K compliment should be 24 but got %d", hll.comp)
	}

	if fmt.Sprintf("%g", hll.alpha) != fmt.Sprintf("%g", 0.7182725932495458) {
		t.Errorf("Alpha should be 0.7182725932495458 but got %g", hll.alpha)
	}
}

func TestNewWithErr(t *testing.T) {
	hll := NewWithErr(0.065)
	if hll.mapSize != 256.0 {
		t.Errorf("Map size should be 256 but got %f", hll.mapSize)
	}

	if len(hll.Table) != int(hll.mapSize) {
		t.Errorf("Table size should be %f but got %d", hll.mapSize, len(hll.Table))
	}

	if hll.comp != 24 {
		t.Errorf("K compliment should be 24 but got %d", hll.comp)
	}

	if fmt.Sprintf("%g", hll.alpha) != fmt.Sprintf("%g", 0.7182725932495458) {
		t.Errorf("Alpha should be 0.7182725932495458 but got %g", hll.alpha)
	}
}

func Test_hashCode(t *testing.T) {
	result := hashCode("apple")
	var expected uint32 = 2297466611
	if result != expected {
		t.Errorf("Hash code for 'apple' should return %d but got %d", expected, result)
	}
}

func TestGetRank(t *testing.T) {
	hll := NewWithErr(0.065)
	hashed := hashCode("kiwi kiwi kiwi")
	rank := getRank(hashed, hll.comp)
	var expected uint32 = 2
	if rank != expected {
		t.Errorf("Rank of 'apple' should be %d but got %d", expected, rank)
	}

	hashed = hashCode("apple")
	rank = getRank(hashed, hll.comp)
	expected = 1
	if rank != expected {
		t.Errorf("Rank of 'apple' should be %d but got %d", expected, rank)
	}
}

func TestAdd(t *testing.T) {
	hll := NewWithErr(0.065)

	hll.Add("apple")
	hashed := hashCode("apple")
	key := uint32(hashed >> hll.comp)
	value := uint32(math.Max(float64(hll.Table[key]), float64(getRank(hashed, hll.comp))))
	var expected uint32 = 1
	if value != expected {
		t.Errorf("Value after add should be %d but got %d", expected, value)
	}

	hll.Add("apple")
	hll.Add("apple")
	value = uint32(math.Max(float64(hll.Table[key]), float64(getRank(hashed, hll.comp))))
	expected = 1
	if value != expected {
		t.Errorf("Value after add should be %d but got %d", expected, value)
	}

	hll.Add("kiwi kiwi kiwi")
	hashed = hashCode("kiwi kiwi kiwi")
	key = uint32(hashed >> hll.comp)
	value = uint32(math.Max(float64(hll.Table[key]), float64(getRank(hashed, hll.comp))))
	expected = 2
	if value != expected {
		t.Errorf("Value after add should be %d but got %d", expected, value)
	}

	hll.Add("apple banana peach wiki pear")
	hashed = hashCode("apple banana peach wiki pear apple")
	key = uint32(hashed >> hll.comp)
	value = uint32(math.Max(float64(hll.Table[key]), float64(getRank(hashed, hll.comp))))
	expected = 4
	if value != expected {
		t.Errorf("Value after add should be %d but got %d", expected, value)
	}
}

func TestCount(t *testing.T) {
	hll := NewWithErr(0.065)
	words := []string{"apple", "this is bananas", "kiwi kiwi kiwi", "Peach is a peach", "apple banana peach wiki pear"}
	for _, word := range words {
		hll.Add(word)
	}
	count := hll.Count()
	if count != uint32(len(words)) {
		t.Errorf("Count should be %d but got %d", len(words), count)
	}
}


hll/error_test.go
package hll

import (
	"fmt"
)

func ExampleHll_Count() {
	hll := NewWithErr(0.065)
	words := []string{"apple", "this is bananas", "kiwi kiwi kiwi", "Peach is a peach", "apple banana peach wiki pear"}
	for _, word := range words {
		hll.Add(word)
	}
	count := hll.Count()
	err := float32((count - uint32(len(words))) / uint32(len(words)) / 100.0)
	fmt.Printf("\nCount: %d\nError: %f%%\n\n", count, err)
	// Output:
	//
	// Count: 5
	// Error: 0.000000%
}


hll/hll_benchmark_test.go
package hll

import (
	"bufio"
	"os"
	"testing"
)

func BenchmarkHllNew(b *testing.B) {
	for i := 0; i < b.N; i++ {
		New()
	}
}

func BenchmarkHllAdd(b *testing.B) {
	hll := New()
	fd, _ := os.Open("/usr/share/dict/words")
	defer fd.Close()

	scanner := bufio.NewScanner(fd)
	var wordCount int
	var totalWords int
	var values []string
	for scanner.Scan() {
		word := scanner.Text()
		totalWords++

		hll.Add(word)
		wordCount++
		values = append(values, word)
	}
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		hll.Add(values[i%totalWords])
	}
}

func BenchmarkHllCount(b *testing.B) {
	hll := New()
	fd, _ := os.Open("/usr/share/dict/words")
	defer fd.Close()

	scanner := bufio.NewScanner(fd)
	for scanner.Scan() {
		hll.Add(scanner.Text())
	}
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		hll.Count()
	}
}


hll/hll.go
package hll

import (
	"math"
)

const (
	// DefaultErr is the default error rate for a hyper-log-log created by
	// using New()
	DefaultErr = 0.065
)

// Hll is the hyper-log-log struct containing the lookup table
// and internal configuration
type Hll struct {
	Table   map[uint32]uint32
	mapSize float64
	comp    uint32
	alpha   float64
	stdErr  float64
}

// New makes a new hyper-log-log with an error rate of DefaultErr
func New() (hll *Hll) {
	hll = &Hll{}
	hll.setupState(DefaultErr)
	hll.setupTable()
	return
}

// NewWithErr takes in an allowed error for the hyper-log-log algorithm
// and will return an Hll
func NewWithErr(err float64) (hll *Hll) {
	hll = &Hll{}
	hll.setupState(err)
	hll.setupTable()
	return
}

func (hll *Hll) setupState(err float64) {
	hll.stdErr = err
	msize := float64(1.04) / hll.stdErr
	k := math.Ceil(math.Log2(msize * msize))
	hll.comp = uint32(32 - k)
	hll.mapSize = math.Pow(2, k)
	hll.alpha = getAlpha(hll.mapSize)
}

func (hll *Hll) setupTable() {
	table := make(map[uint32]uint32)
	var i uint32
	for ; i < uint32(hll.mapSize); i++ {
		table[i] = 0
	}
	hll.Table = table
}

func getAlpha(msize float64) (alpha float64) {
	if msize == 16 {
		alpha = 0.673
	} else if msize == 32 {
		alpha = 0.697
	} else if msize == 64 {
		alpha = 0.709
	} else {
		alpha = 0.7213 / (1.0 + 1.079/msize)
	}
	return
}

func hashCode(key string) (hashed uint32) {
	var i uint32
	runes := []rune(key)

	for ; i < uint32(len(runes)); i++ {
		hashed += uint32(runes[i])
		hashed += (hashed << 10)
		hashed ^= (hashed >> 6)
	}

	hashed += (hashed << 3)
	hashed ^= (hashed >> 11)
	hashed += (hashed << 15)

	return
}

func getRank(hashed uint32, max uint32) (rank uint32) {
	var one uint32 = 1
	for rank = 1; (hashed&one) == 0 && rank <= max; rank++ {
		hashed = hashed >> 1
	}
	return rank
}

// Add takes a term to add to the hyper-log-log table
// and will be hashed and incremented
func (hll *Hll) Add(term string) {
	hashed := hashCode(term)
	key := uint32(hashed >> hll.comp)

	current := float64(hll.Table[key])
	rank := float64(getRank(hashed, hll.comp))
	hll.Table[key] = uint32(math.Max(current, rank))
}

// Count will calculate the size of the terms in the hyper-log-log table
// and return the total, consider that there may be false positives.
func (hll *Hll) Count() uint32 {
	var c float64
	var i uint32

	for ; i < uint32(hll.mapSize); i++ {
		c += 1.0 / math.Pow(2, float64(hll.Table[i]))
	}

	e := hll.alpha * hll.mapSize * hll.mapSize / c

	// Make corrections
	if e <= (5/2)*hll.mapSize {
		var v float64
		for i = 0; i < uint32(hll.mapSize); i++ {
			if hll.Table[i] == 0 {
				v++
			}
		}
		if v > 0 {
			e = hll.mapSize * math.Log(hll.mapSize/v)
		}
	} else {
		pow32 := math.Pow(2, 32)
		if e > (1.0/30.0)*pow32 {
			e = -pow32 * math.Log(1.0-e/pow32)
		}
	}
	return uint32(e)
}
